week-1-ts-ecosystem/w01e03-object-oriented-structure/src/shop/cart.class.ts

```diff
@@ -0,0 +1,72 @@
+import {FreeProductInterface, ProductInterface} from "./product.interface";
+import {CartType} from "./cart.type";
+
+
+export default class Cart {
+    #products: (ProductInterface | FreeProductInterface)[] = [];
+    #type: CartType;
+    #sum = 0;
+
+    constructor(type: CartType) {
+        this.#type = type;
+    }
+
+    addItem(product: ProductInterface | FreeProductInterface): void {
+        const existingProduct = this.#products.findIndex(item => item.id === product.id);
+
+        if (existingProduct > -1) {
+            this.updateCart(product.id);
+            return;
+        }
+
+        if (product.type === this.#type) {
```

Comment:
Niestety, ponieważ rozróżniasz swoje elementy koszyka po wartości `type` - to podpisałeś na siebie:
"klauzulę o `if`ologii stosowanej" - z której się nie wygrzebiesz, jeśli będą dodatkowe różniące się od siebie produkty

Tam w linii `:6` zamiast Unii typów - powinien być typ generyczny, żebyśmy mogli się tego pozbyć